src/main/java/com/aws/greengrass/deployment/ConnectivityValidator.java

```diff
@@ -0,0 +1,224 @@
+    }
+
+    private boolean greengrassDataPlanePortHasChanged(DeviceConfiguration currentdeviceConfiguration) {
+        int currentGreengrassDataPlanePort = Coerce.toInt(currentdeviceConfiguration
+                .getGreengrassDataPlanePort().getOnce());
+        int newGreengrassDataPlanePort = Coerce.toInt(deviceConfiguration.getGreengrassDataPlanePort().getOnce());
+        return currentGreengrassDataPlanePort != newGreengrassDataPlanePort;
+    }
+
+    /**
+     * Creates an MQTT client and checks if the device can connect to AWS.
+     *
+     * @param totallyCompleteFuture Future will be updated if validation fails
+     * @throws InterruptedException if sleep is interrupted
+     */
+    public boolean mqttClientCanConnect(CompletableFuture<DeploymentResult> totallyCompleteFuture)
+            throws InterruptedException {
+        logger.atDebug().log("Checking MQTT client can connect");
+
+        builder = mqttClient.createMqttConnectionBuilder(deviceConfiguration, null);
+        String clientId = Coerce.toString(deviceConfiguration.getThingName()) + "#validation";
+        connection = builder.withClientId(clientId).build();
+
+        for (int i = 0; i < maxRetries; i++) {
+            try {
+                connection.connect()
+                        .get(mqttClient.getMqttOperationTimeoutMillis(deviceConfiguration), TimeUnit.MILLISECONDS);
+                return true;
+            } catch (MqttException | TimeoutException | ExecutionException e) {
+                Thread.sleep(TimeUnit.SECONDS.toMillis(2));
+            }
+        }
+        totallyCompleteFuture
+                .complete(new DeploymentResult(DeploymentResult.DeploymentStatus.FAILED_NO_STATE_CHANGE,
+                        new ComponentConfigurationValidationException(
+                        "MQTT client failed to connect to IoT Core using new configurations",
+                        DeploymentErrorCode.FAILED_TO_RECONNECT)));
+        return false;
+    }
+
+    /**
+     * Creates an HTTP client and checks if the device can connect to AWS.
+     *
+     * @param totallyCompleteFuture Future will be updated if validation fails
+     * @throws InterruptedException if sleep is interrupted
+     */
+    public boolean httpClientCanConnect(CompletableFuture<DeploymentResult> totallyCompleteFuture)
+            throws InterruptedException {
+        logger.atDebug().log("Checking HTTP client can connect");
+
+        if (factory == null) {
+            factory = new GreengrassServiceClientFactory(deviceConfiguration);
+        }
+        try {
+            greengrassV2DataClient = factory.fetchGreengrassV2DataClient();
+        } catch (DeviceConfigurationException e) {
+            totallyCompleteFuture
+                    .complete(new DeploymentResult(DeploymentResult.DeploymentStatus.FAILED_NO_STATE_CHANGE,
+                            new ComponentConfigurationValidationException(
```

Comment:
same here, I think you should append the validation error in the exception message

Reply:
The exception messages weren't really actionable, I currently think it's better if we just log those messages instead